docs/reference/mapping/types/dense-vector.asciidoc

```diff
@@ -45,5 +45,137 @@ PUT my-index-000001/_doc/2
+        "dims": 3,
+        "index": true,
+        "similarity": "dot_product" <1>
+      }
+    }
+  }
+}
+--------------------------------------------------
+<1> When `index` is enabled, you must define the vector similarity to use in kNN search
+
+Elasticsearch uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to
+support efficient kNN search. Like most kNN algorithms, HNSW is an approximate
+method that sacrifices result accuracy for improved speed.
+
+NOTE: indexing vectors for kNN search is an expensive process. It can take
+substantial time to ingest documents that contain vector fields with `index`
+enabled.
+
+[discrete]
+[[dense-vector-params]]
+==== Parameters for dense vector fields
+
+The following mapping parameters are accepted:
+
+`dims`::
+
+The number of vector dimensions. This value is required and cannot exceed 2048.
+
+`index`::
+
+Whether the vectors should be indexed to support efficient k-nearest neighbor search. Defaults to `false`.
+
+`similarity`::
+
+The vector similarity metric to use in kNN search. Documents are ranked by
+their vector field's similarity to the query vector. The `_score` of each
+document will be derived from the similarity, in a way that ensures scores are
+positive and that a larger score corresponds to a higher ranking.
++
+When `index` is enabled, this parameter is required. The following options are supported:
+
+`l2_norm`:::
+Computes similarity based on the L^2^ distance (also known as Euclidean
+distance) between the vectors. The document `_score` is computed as
+`1 / (1 + l2_norm(query, vector)^2)`.`
+
+`dot_product`:::
+Computes the dot product of two vectors. This option provides an optimized way
+to perform cosine similarity. In order to use it, all vectors must be of unit
+length, including both document and query vectors. The document `_score` is
+computed as `(1 + dot_product(query, vector)) / 2`.
+
+`cosine`:::
+Computes the cosine similarity. Note that the most efficient way to perform
+cosine similarity is to normalize all vectors to unit length, and instead use
+`dot_product`. You should only use `cosine` if you need to preserve the
+original vectors and cannot normalize them in advance. The document `_score`
+is computed as `(1 + dot_product(query, vector)) / 2`.
+
+NOTE: Although they are conceptually related, the `similarity` parameter is
+different from <<text,`text`>> field <<similarity,`similarity`>> and accepts
+a distinct set of options.
+
+`index_options`::
+
+An optional section that configures the kNN indexing algorithm. The HNSW
+algorithm has two internal parameters that influence how the data structure is
+built. These can be adjusted to improve the accuracy of results, at the
+expense of slower indexing speed. The `index_options` object contains the
+following properties:
+
+`type`:::
+The type of kNN algorithm to use. Currently only `hnsw` is supported.
+
+`m`:::
+Controls how many of the nearest neighbor candidates are connected to each new
```

Comment:
may be better say, the maximum number of neighbour connections each node can have in the HNSW graph?

Reply:
👍  I reworded this to make it clearer. But I didn't say "maximum" because this is the actual number of connections that the algorithm tries to create (not an upper bound).

Reply:
My interpretation was that it is an upper bound, in the sense that we don't allow to have more neighbours than `m`, and will prune the worst neighbours to have the number of neighbours always under `m`.

Reply:
Oh I see, I interpreted "maximum" to mean that it's possible many nodes have fewer than `m` neighbors. But the algorithm attempts to have each node have exactly `m` neighbors.

In any case, hopefully the current wording is clear :) It matches the [description from hnswlib](https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md): "M - the number of bi-directional links created for every new element during construction".